### PR TITLE
[SPARK-48849][SS]Create OperatorStateMetadataV2 for the TransformWithStateExec operator

### PR DESCRIPTION
**Pull request by ericm-db**
### What changes were proposed in this pull request?

Introducing the OperatorStateMetadataV2 format that integrates with the TransformWithStateExec operator. This is used to keep information about the TWS operator, will be used to enforce invariants in between query runs. Each OperatorStateMetadataV2 has a pointer to the StateSchemaV3 file for the corresponding operator.
Will introduce purging in this PR: https://github.com/apache/spark/pull/47286
### Why are the changes needed?

This is needed for State Metadata integration with the TransformWithState operator. 

### How was this patch tested?

Added unit tests to StateStoreSuite and TransformWithStateSuite

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 1]
```diff
@@ -66,7 +67,7 @@ class StateSchemaCompatibilityChecker(
     try {
       val versionStr = inStream.readUTF()
       // Ensure that version 3 format has schema file path provided explicitly
-      if (versionStr == "v3" && schemaFilePath.isEmpty) {
+      if (versionStr == "v3" && oldSchemaFilePath.isEmpty) {
```
Comment (jingz-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:70]: IIUC, is this supposed to be `newSchemaFilePath`?

Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:70]: Ah, we don't actually need this. Will remove

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala [hunk 2]
```diff
@@ -188,29 +191,56 @@ class StateMetadataPartitionReader(
...
-  private def allOperatorStateMetadata: Array[OperatorStateMetadata] = {
+  private[sql] def allOperatorStateMetadata: Array[(OperatorStateMetadata, Long)] = {
     val stateDir = new Path(checkpointLocation, "state")
     val opIds = fileManager
       .list(stateDir, pathNameCanBeParsedAsLongFilter).map(f => pathToLong(f.getPath)).sorted
-    opIds.map { opId =>
-      new OperatorStateMetadataReader(new Path(stateDir, opId.toString), hadoopConf).read()
+    opIds.flatMap { opId =>
+      val operatorIdPath = new Path(stateDir, opId.toString)
+      // check if OperatorStateMetadataV2 path exists, if it does, read it
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala:200]: We can already read the version here right ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala [hunk 3]
```diff
@@ -188,29 +191,56 @@ class StateMetadataPartitionReader(
...
+        case v2: OperatorStateMetadataV2 =>
+          v2.stateStoreInfo.map { stateStoreMetadata =>
+            StateMetadataTableEntry(v2.operatorInfo.operatorId,
+              v2.operatorInfo.operatorName,
+              stateStoreMetadata.storeName,
+              stateStoreMetadata.numPartitions,
+              batchId,
+              batchId,
+              v2.operatorPropertiesJson,
+              stateStoreMetadata.numColsPrefixKey
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala:240]: For v2 - this is kind of misleading. Could we just remove this ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala [hunk 4]
```diff
@@ -325,6 +340,19 @@ class HDFSMetadataLog[T <: AnyRef : ClassTag](sparkSession: SparkSession, path:
     }
   }
 
+  def purgeOldest(minEntriesToMaintain: Int): Unit = {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:343]: Lets move the purging out to a separate PR ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OperatorStateMetadataLog.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  def this(sparkSession: SparkSession, path: String) = {
+    this(
+      sparkSession.sessionState.newHadoopConf(),
+      path,
+      metadataCacheEnabled = sparkSession.sessionState.conf.getConf(
+        SQLConf.STREAMING_METADATA_CACHE_ENABLED)
+    )
+  }
+
+  override protected def serialize(metadata: OperatorStateMetadata, out: OutputStream): Unit = {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OperatorStateMetadataLog.scala:47]: Better to consolidate this with the existing implementation for the v1 format

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala [hunk 6]
```diff
@@ -188,29 +191,56 @@ class StateMetadataPartitionReader(
...
+            )
+          }
+        case v2: OperatorStateMetadataV2 =>
+          v2.stateStoreInfo.map { stateStoreMetadata =>
+            StateMetadataTableEntry(v2.operatorInfo.operatorId,
+              v2.operatorInfo.operatorName,
+              stateStoreMetadata.storeName,
+              stateStoreMetadata.numPartitions,
+              batchId,
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala:237]: We can just retain values as v1 here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 7]
```diff
@@ -219,7 +222,8 @@ object StateSchemaCompatibilityChecker {
       stateSchemaVersion: Int,
       extraOptions: Map[String, String] = Map.empty,
       storeName: String = StateStoreId.DEFAULT_STORE_NAME,
-      schemaFilePath: Option[Path] = None): StateSchemaValidationResult = {
+      oldSchemaFilePath: Option[Path] = None,
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:225]: Lets update the comments above ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 8]
```diff
@@ -185,7 +188,7 @@ class StateSchemaCompatibilityChecker(
     new Path(new Path(storeCpLocation, "_metadata"), "schema")
 }
 
-object StateSchemaCompatibilityChecker {
+object StateSchemaCompatibilityChecker extends Logging {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:191]: Do we need to extend here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala [hunk 9]
```diff
@@ -208,13 +208,24 @@ class IncrementalExecution(
...
-          case stateStoreWriter: StateStoreWriter =>
-            val metadata = stateStoreWriter.operatorStateMetadata()
-            // TODO: [SPARK-48849] Populate metadata with stateSchemaPaths if metadata version is v2
-            val metadataWriter = new OperatorStateMetadataWriter(new Path(
-              checkpointLocation, stateStoreWriter.getStateInfo.operatorId.toString), hadoopConf)
+          case ssw: StateStoreWriter =>
+            val metadata = ssw.operatorStateMetadata(stateSchemaPaths)
+            val metadataWriter = ssw.operatorStateMetadataVersion match {
+              case 1 =>
+                new OperatorStateMetadataV1Writer(new Path(
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala:218]: Maybe rephrase as

```
new OperatorStateMetadataWriter(checkpointLoc, opId, hadoopConf, operatorMetadataVersion, Some(currentBatchId))
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala [hunk 10]
```diff
@@ -382,12 +388,46 @@ case class TransformWithStateExec(
...
+    val newSchemas = getColFamilySchemas()
     val stateSchemaDir = stateSchemaDirPath()
-    val stateSchemaFilePath = new Path(stateSchemaDir, s"${batchId}_${UUID.randomUUID().toString}")
-    List(StateSchemaCompatibilityChecker.validateAndMaybeEvolveStateSchema(getStateInfo, hadoopConf,
-      newColumnFamilySchemas.values.toList, session.sessionState, stateSchemaVersion,
-      schemaFilePath = Some(stateSchemaFilePath)))
+    val newStateSchemaFilePath =
+      new Path(stateSchemaDir, s"${batchId}_${UUID.randomUUID().toString}")
+    val metadataPath = new Path(getStateInfo.checkpointLocation, s"${getStateInfo.operatorId}")
+    val metadataReader = new OperatorStateMetadataV2Reader(metadataPath, hadoopConf)
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:396]: We can just pass the required version and get the correct reader ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala [hunk 11]
```diff
@@ -86,24 +123,96 @@ object OperatorStateMetadataUtils {
...
+  private implicit val manifest = Manifest
+    .classType[OperatorStateMetadataV2](implicitly[ClassTag[OperatorStateMetadataV2]].runtimeClass)
+
+  def metadataDirPath(stateCheckpointPath: Path): Path =
+    new Path(new Path(new Path(stateCheckpointPath, "_metadata"), "metadata"), "v2")
+
+  def metadataFilePath(stateCheckpointPath: Path, currentBatchId: Long): Path =
+    new Path(metadataDirPath(stateCheckpointPath), currentBatchId.toString)
+
+  def deserialize(in: BufferedReader): OperatorStateMetadata = {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala:191]: Where do we use this ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala [hunk 12]
```diff
@@ -124,22 +233,96 @@ class OperatorStateMetadataWriter(stateCheckpointPath: Path, hadoopConf: Configu
...
       val version = MetadataVersionUtil.validateVersion(versionStr, 1)
-      OperatorStateMetadataUtils.deserialize(version, inputReader)
+      Some(OperatorStateMetadataUtils.deserialize(version, inputReader))
+    } finally {
+      inputStream.close()
+    }
+  }
+}
+
+class OperatorStateMetadataV2Writer(
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala:263]: Isn't this redundant here ? could we combine the v1/v2 interfaces here ?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala [hunk 13]
```diff
@@ -63,6 +63,33 @@ class RunningCountStatefulProcessor extends StatefulProcessor[String, String, (S
...
+      _countState.clear()
+      Iterator.empty
+    } else {
+      _countState.update(count)
+      Iterator((key, count.toString))
+    }
+  }
+}
+
+
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala:92]: nit: extra newline ?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala [hunk 14]
```diff
@@ -63,6 +63,33 @@ class RunningCountStatefulProcessor extends StatefulProcessor[String, String, (S
   }
 }
 
+class RunningCountStatefulProcessorInt extends StatefulProcessor[String, String, (String, String)]
+  with Logging {
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala:67]: Do we need `Logging` here ?

**Review by anishshri-db — APPROVED** [review]

**Comment by ericm-db** [comment]
@HeartSaVioR PTAL, thanks!
=

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 15]
```diff
@@ -98,7 +95,7 @@ class StateSchemaCompatibilityChecker(
       stateStoreColFamilySchema: List[StateStoreColFamilySchema],
       stateSchemaVersion: Int): Unit = {
     // Ensure that schema file path is passed explicitly for schema version 3
-    if (stateSchemaVersion == 3 && schemaFilePath.isEmpty) {
+    if (stateSchemaVersion == 3 && newSchemaFilePath.isEmpty) {
       throw new IllegalStateException("Schema file path is required for schema version 3")
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:99]: Let's apply error class framework here while we are here. internalError would be OK.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 16]
```diff
@@ -208,7 +211,10 @@ object StateSchemaCompatibilityChecker {
    * @param stateSchemaVersion - version of the state schema to be used
    * @param extraOptions - any extra options to be passed for StateStoreConf creation
    * @param storeName - optional state store name
-   * @param schemaFilePath - optional schema file path
+   * @param oldSchemaFilePath - optional path to the old schema file. If not provided, will default
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:214]: Is it valid for schema version 3 to have None? Otherwise let's also mention here as same as below; Needed for schema version 3.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala [hunk 17]
```diff
@@ -60,19 +81,65 @@ case class OperatorStateMetadataV1(
...
+    stateStoreInfo: Array[StateStoreMetadataV2],
+    operatorPropertiesJson: String) extends OperatorStateMetadata {
+  override def version: Int = 2
+}
+
+object OperatorStateMetadataUtils extends Logging {
+
+  sealed trait OperatorStateMetadataReader {
+    def version: Int
+
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala:95]: nit: shall we be consistent? see below trait - no empty line between two or both empty line

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala [hunk 18]
```diff
@@ -60,19 +81,65 @@ case class OperatorStateMetadataV1(
...
+      val versionStr = inputReader.readLine()
+      val version = MetadataVersionUtil.validateVersion(versionStr, 2)
+      Some(deserialize(version, inputReader))
+    } finally {
+      inputStream.close()
+    }
+  }
+
+  def writeMetadata(
+    outputStream: CancellableFSDataOutputStream,
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala:119]: nit: 2 more spaces

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala [hunk 19]
```diff
@@ -86,62 +153,162 @@ object OperatorStateMetadataUtils {
...
-    } finally {
-      inputStream.close()
+    OperatorStateMetadataUtils.readMetadata(inputStream)
+  }
+}
+
+class OperatorStateMetadataV2Writer(
+    stateCheckpointPath: Path,
+    hadoopConf: Configuration,
+    currentBatchId: Long) extends OperatorStateMetadataWriter with Logging {
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala:268]: nit: Logging does not seem to be used

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala [hunk 20]
```diff
@@ -86,62 +153,162 @@ object OperatorStateMetadataUtils {
...
+
+    fm.mkdirs(metadataFilePath.getParent)
+    val outputStream = fm.createAtomic(metadataFilePath, overwriteIfPossible = false)
+    OperatorStateMetadataUtils.writeMetadata(outputStream, operatorMetadata, metadataFilePath)
+  }
+}
+
+class OperatorStateMetadataV2Reader(
+    stateCheckpointPath: Path,
+    hadoopConf: Configuration) extends OperatorStateMetadataReader with Logging {
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala:288]: ditto, Logging does not seem to be used

**Review by HeartSaVioR — COMMENTED** [review]
First pass. Looks good in general.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala (hunk 15 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:99]: I see lots of places to do this; let's postpone.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala (hunk 16 above)
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:214]: Yeah, we don't expect there to be a schema file if we are in the first run of a new query, right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala [hunk 21]
```diff
@@ -208,13 +208,16 @@ class IncrementalExecution(
...
+        val stateSchemaPaths = schemaValidationResult.map(_.schemaPath)
         // write out the state schema paths to the metadata file
         statefulOp match {
-          case stateStoreWriter: StateStoreWriter =>
-            val metadata = stateStoreWriter.operatorStateMetadata()
-            // TODO: [SPARK-48849] Populate metadata with stateSchemaPaths if metadata version is v2
-            val metadataWriter = new OperatorStateMetadataWriter(new Path(
-              checkpointLocation, stateStoreWriter.getStateInfo.operatorId.toString), hadoopConf)
+          case ssw: StateStoreWriter =>
+            val metadata = ssw.operatorStateMetadata(stateSchemaPaths)
```
[the remark this answers is not in the document]
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala:215]: Yeah, even if there is no change, we will still write a new file.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala [hunk 22]
```diff
@@ -188,29 +191,54 @@ class StateMetadataPartitionReader(
...
   }
 
-  private def allOperatorStateMetadata: Array[OperatorStateMetadata] = {
+  private[sql] def allOperatorStateMetadata: Array[OperatorStateMetadata] = {
     val stateDir = new Path(checkpointLocation, "state")
     val opIds = fileManager
       .list(stateDir, pathNameCanBeParsedAsLongFilter).map(f => pathToLong(f.getPath)).sorted
-    opIds.map { opId =>
-      new OperatorStateMetadataReader(new Path(stateDir, opId.toString), hadoopConf).read()
+    opIds.flatMap { opId =>
```
[the remark this answers is not in the document]
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala:198]: I see, is this the case in which we have added an operator after the first batch of the query? Like what we expect to happen?
I'm not sure what the behavior should be in this case. I think just 'dropping' the operator makes sense, but I can also conform to the previous interface, and fail the query if the operator does not have any metadata. What do you think @anishshri-db ?

[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala:199]: dropping the operator will not tell the truth for users; discussed offline.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala [hunk 23]
```diff
@@ -17,9 +17,7 @@
 package org.apache.spark.sql.execution.datasources.v2.state.metadata
 
 import java.util
-
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala:20]: @ericm-db - we need to revert this back to original ?

Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala]: Sorry, inadvertent.

**Comment by ericm-db** [comment]
@HeartSaVioR addressed feedback, please take another look when you get a chance

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.
